### PR TITLE
Feature/red 6 fhir payload

### PR DESCRIPTION
**Pull request by alibombaye**
Adaptor that takes in a fhir payload for a birth registration and generates an edifact message.

To run the sample script `pipenv run transfer` which takes the sample json payload from the GP mailbox. Runs the adaptor and creates an edifact file in the NHAIS inbox.

### CODE REVIEW

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        :param first_name: first name of the patient
+        :param last_name: last name of the patient
+        :param gender: patient gender
+        :param date_of_birth: date of birth of patient
+        :param address_line_1: first line of address
+        :param city: city of patient
+        :param postcode: post code of patient
+        :param place_of_birth: place of birth of patient
+        :return: Patient
+        """
```
Comment (Philip-Woods) [inline nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: As discussed, we suggested remove trivial documentation and try to keep to docs that explain an unclear variable name, return type or action.

Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: I agree, the documentation here seems unnecessary 👍

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        Create the edifact interchange from the fhir operation definition
+        :param fhir_operation:
+        :return: Interchange
+        """
+        interchange_sequence_number = odh.get_parameter_value(fhir_operation,
+                                                              parameter_name=ParameterName.INTERCHANGE_SEQ_NO)
+        sender_cypher = odh.get_parameter_value(fhir_operation, parameter_name=ParameterName.SENDER_CYPHER)
+        nhais_cypher = odh.get_parameter_value(fhir_operation, parameter_name=ParameterName.NHAIS_CYPHER)
+        recipient = ''
+        if len(nhais_cypher) == 3:
```
Comment (gareth-allan) [inline nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: This is a little opaque. What's happening here? Probably worth a comment, or extracting to a nicely-named method.

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            recipient = nhais_cypher + '1'
+        elif len(nhais_cypher) == 2:
+            recipient = nhais_cypher + "01"
+
+        messages = Messages(messages=[MessageAdaptor.create_message(fhir_operation)])
+
+        interchange = Interchange(sender=sender_cypher, recipient=recipient,
+                                  sequence_number=interchange_sequence_number,
+                                  date_time=fhir_operation.date.as_json(), messages=messages).to_edifact()
```
Comment (gareth-allan) [inline nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: I think it's worth extracting the `to_edifact()` call to a separate line, as it's easy to miss it on the end of the `Interchange()` call.

File: nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+class InterchangeAdaptorTest(unittest.TestCase):
+    """
+    Test the conversation of fhir to an edifact interchange
+    """
+
+    def test_create_interchange(self):
+        """
+        Test the function to create an edifact interchange
+        """
+        with self.subTest("When the operation is for a Birth Registration"):
```
Comment (gareth-allan) [inline nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py]: Again, I'm not sure of the value of having a single subtest. Is the goal to provide a friendly description for the test? Couldn't you do that with a message when you make an assertion?

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/fixtures.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+from adaptor.outgoing.fhir_helpers.operation_definition import OperationDefinitionHelper as odh
```
Comment (gareth-allan) [inline nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/fixtures.py]: I'm not sure this module makes sense in `fhir_helpers/tests`, as far as I can see, it's only used in `/adaptor/outgoing/tests`.

File: nhais-adaptor/adaptor/outgoing/message_adaptor.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    @staticmethod
+    def create_message_beginning(fhir_operation):
+        """
+        Create the beginning of the message
+        :param fhir_operation:
+        :return: MessageBeginning
+        """
+        nhais_id = odh.get_parameter_value(fhir_operation=fhir_operation, parameter_name=ParameterName.NHAIS_CYPHER)
+
+        if fhir_operation.name == OperationName.REGISTER_BIRTH:
```
Comment (gareth-allan) [inline nhais-adaptor/adaptor/outgoing/message_adaptor.py]: This doesn't look like it'll be very flexible when we look to expand to supporting other operations. Is this class specific to the birth registration message?

File: nhais-adaptor/Pipfile [hunk 7]
```diff
@@ -12,4 +13,5 @@ fhirclient = "*"
 python_version = "3.7"
 
 [scripts]
-tests = "python -m unittest discover -p '*_test.py' -v"
+tests = "python -m unittest discover -p 'test_*.py' -v"
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/Pipfile]: 👍 yeah no longer needed. Updated to remove the pattern matching and the discover arg too

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+from fhirclient.models.identifier import Identifier
+from fhirclient.models.operationdefinition import OperationDefinitionParameterBinding, OperationDefinitionParameter, \
+    OperationDefinition
+from fhirclient.models.patient import Patient
+from fhirclient.models.practitioner import Practitioner
+from fhirclient.models.address import Address
+from fhirclient.models.extension import Extension
+
+
+class OperationDefinitionHelper:
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: Yes you are right. I suppose it's a FhirHelper really. To assist in anything Fhir related. Does `FhirHelper` sound like a more sensible name?

File: nhais-adaptor/tests/test_fhir_to_edifact.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        Test when the operation is for a birth registration
+        """
+        with open("./tests/edifact.txt", "r") as expected_file:
+            expected_edifact_interchange = expected_file.read()
+
+        with open("./tests/patient-register-birth.json", "r") as patient_register:
+            patient_register_json = json.load(patient_register)
+        op_def = OperationDefinition(patient_register_json)
+        edifact_interchange = InterchangeAdaptor.create_interchange(fhir_operation=op_def)
+        generated_edifact_interchange = "'\n".join(edifact_interchange.split("'"))
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/tests/test_fhir_to_edifact.py]: Done 👍

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+    @staticmethod
+    def create_operation_definition(name, code, date_time, contained, parameter):
+        """
+        Helper function to create a fhir operation definition
+        :param name: The name of the operation
+        :param code: The code of the operation
+        :param date_time: the time stamp of the operation
+        :param contained: list of contained resources
+        :param parameter: list of parameters
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: I wanted to keep it close to the Fhir models as possible to avoid confusion. Since it is called `parameter` in the model even though it is a list. However, I suppose this helper function abstracts some of that fhir craziness so will change to `parameters`

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+                           'identifier': [nhs_number.as_json()],
+                           'gender': gender,
+                           'name': [name.as_json()],
+                           'birthDate': date_of_birth,
+                           'address': [pat_address.as_json()],
+                           'extension': [place_of_birth.as_json()]})
+        return patient
+
+    @staticmethod
+    def get_parameter_value(fhir_operation, parameter_name):
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: That's correct. Will extract out to a separate module

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        op = odh.create_operation_definition(name="some.name",
+                                             code="some.code",
+                                             date_time="2019-04-23 09:00:04.159338",
+                                             contained=[],
+                                             parameter=[op_param_transaction_number])
+
+        self.assertEqual(odh.get_parameter_value(op, "some_param_name"), "17")
+
+    def test_find_resource(self):
+        with self.subTest("When practitioner details are found in the operation definition"):
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py]: I find it bit more descriptive than just a test function name. Also I was thinking it could have additional test for edge cases like not finding a resource or finding multiples that I haven't catered for at the moment.

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                                                     parameter=[])
+
+            self.assertIsInstance(op_def, OperationDefinition)
+            self.assertEqual(op_def.name, "name.of.operation")
+            self.assertEqual(op_def.code, "code.of.operation")
+            self.assertEqual(op_def.date.as_json(), "2019-04-23 09:00:04.159338")
+            self.assertEqual(op_def.contained, None)
+            self.assertEqual(op_def.parameter, None)
+
+        with self.subTest("with one parameter"):
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py]: 👍 Good shout. Will add a subTest for this scenario

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        self.assertEqual(practitioner.identifier[0].value, "1234567")
+        self.assertEqual(practitioner.identifier[1].value, "111")
+
+    def test_create_patient(self):
+        patient = odh.create_patient_resource(resource_id="patient-1", nhs_number="NHSNO11111",
+                                              title="Mr", first_name="Peter", last_name="Parker",
+                                              gender="male", date_of_birth="2019-04-23", place_of_birth="Spidey Town",
+                                              address_line_1="1 Spidey Way", city="Spidey Town", postcode="SP1 1AA")
+
+        self.assertIsInstance(patient, Patient)
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/test_operation_definition.py]: yeah bit of laziness on my part due to the sheer amount of parameters to assert on :(

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py (hunk 2 above)
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: 👍 done

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py (hunk 3 above)
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: 👍 done

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: renamed and separate out fhir_creators and fhir_finders

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/operation_definition.py]: separate module created

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/fixtures.py (hunk 5 above)
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/tests/fixtures.py]: 👍 Done

File: nhais-adaptor/adaptor/outgoing/message_adaptor.py (hunk 6 above)
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/message_adaptor.py]: I've added a dictionary to map these codes based upon the operation name. Should make this more flexible going forward

File: nhais-adaptor/adaptor/outgoing/fhir_helpers/constants.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
+class OperationName:
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/fhir_helpers/constants.py]: good shout I have moved these to a more relevant place

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py [hunk 16]
```diff
@@ -1,35 +1,47 @@
...
+An adaptor to take in fhir models and generate an edifact interchange
+"""
 
-class InterchangeAdaptor:
+
+def generate_recipient_from(nhais_cypher):
     """
-    An adaptor to take in fhir models and generate an edifact interchange
+    Generate the recipient cypher.
+    Will append '1' when the nhais cypher is 3 characters and '01' if 2 characters
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: This is an edifact requirement and something that was explained to me by Mark;
`The “XX1” value if the Cipher of the NHAIS system. This can actually be 2 or 3 alpha characters. For example, “BD” is the Cipher for the Bedfordshire NHAIS System; “BRS” is the Cipher for the Avon NHAIS System. The “XX11” value is the EDI Trading Partner for the NHAIS system. This is the NHAIS system cipher with either “01” added on the end (for 2 character ciphers) or “1” added on the end (for 3 character ciphers). So, the Trading Partner for the Bedfordshire NHAIS system (BD) is “BD01” and the Trading Partner for the Avon NHAIS system (BRS) is “BRS1”.`

Basically we are able to deduce the recipient cypher from the nhais cypher. 
Do you think the function can be named a little better to explain this or add more detail in the docstring?

File: nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                                            "LOC+950+SPIDEY TOWN'"
+                                            "S02+2'"
+                                            "PNA+PAT+NHSNO11111:OPI+++SU:PARKER+FO:PETER+TI:MR'"
+                                            "DTM+329:20190420:102'"
+                                            "PDI+1'"
+                                            "NAD+PAT++:1 SPIDEY WAY::SPIDEY TOWN:+++++SP1 1AA'"
+                                            "UNT+18+000001'"
+                                            "UNZ+1+000001'")
+
+            op_param_interchange_sequence = creators.create_parameter_with_binding(
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py]: Yeah I think that would work nicely. I'll add something to the fixtures for this 👍

File: nhais-adaptor/adaptor/outgoing/interchange_adaptor.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (gareth-allan) [reply nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: I'd say a little more detail in the docstring to explain what we're doing/how we know what to do would make sense. This strikes me as one of those things somebody will come to and have no idea why it's being done, so it would be good to capture some of the above in the comments.

I think the function name makes sense as it is.

[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/interchange_adaptor.py]: 👍 Done

File: nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/tests/test_interchange_adaptor.py]: 👍 Done

File: nhais-adaptor/adaptor/outgoing/message_adaptor.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    return edi_name
+
+
+def create_patient_address(fhir_patient_address):
+    """
+    Function to generate the edifact representation of the patient address
+    :param fhir_patient_address: The fhir representation of the patient address
+    :return: edifact address
+    """
+    address_lines = ["", "", ""]
```
[the remark this answers is not in the document]
Comment (alibombaye) [reply nhais-adaptor/adaptor/outgoing/message_adaptor.py]: Good shout. I've extracted this logic out to a separate function with some documentation and tests that should explain what it does. Let me know if still isn't clear
